### PR TITLE
Add documentation for unit.py unit conversion feature

### LINKED ISSUE #2512: Add documentation about unit conversion feature
## Enhancement proposal
This feature was implemented in https://github.com/ManimCommunity/manim/pull/742, but there isn't any documentation about this. 
```py
# 50px -> Munit
>>> 50 * unit.Pixels
0.37037037037037035
# 90 degrees to radians
>>> 90 * unit.Degrees
1.5707963267948966
# 10 % of X-axis
>>> unit.Percent(X_AXIS) * 10
1.4222222222222223
```

### PR DESCRIPTION
**Pull request by tameTNT**
## Overview: What does this pull request change?

- Adds documentation and 2 examples for unit helper classes `Percent`, `Pixels` and `Degrees`
- (Hopefully) Fixes bug in `class.rst` where *Methods* header is displayed even when there are no public methods to list.

## Motivation and Explanation: Why and how do your changes improve the library?

- Closes #2512 

## Links to added or changed documentation pages

- https://manimce--2604.org.readthedocs.build/en/2604/reference_index/utilities_misc.html
- https://manimce--2604.org.readthedocs.build/en/2604/reference/manim.utils.unit.html#module-manim.utils.unit
- https://manimce--2604.org.readthedocs.build/en/2604/reference/manim.utils.unit.Percent.html#manim.utils.unit.Percent
- https://docs.manim.community/en/stable/reference/manim.utils.tex.TexTemplate.html#:~:text=str-,Methods,-add_to_document (example of potentially affected methods header in current docs)
- https://manimce--2604.org.readthedocs.build/en/2604/reference/manim.utils.tex.TexTemplate.html#:~:text=str-,Methods (with proposed changes - hopefully no different)

## Further Information and Comments

- This *is* my first contribution to the repository! I tried to follow all the style guides and documentation formats given by the contributing wiki but I've inevitably missed something I'm sure. Just let me know where I've messed up and I'll try and fix it right away.
- I have a habit of making lots of small commits instead of fewer larger ones so please **squash and merge** this PR instead of just merging it for everyone's sake!

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by tameTNT** [comment]
And I've found my first mistake - I didn't actually update the docs.
Apologies if this is an incredibly stupid question but how do I actually do that?
Do I just add `~utils.unit` to `docs/source/reference_index/utilities_misc.rst`?

**Comment by tameTNT** [comment]
> Do I just add `~utils.unit` to `docs/source/reference_index/utilities_misc.rst`?
Okay so I tried this and it worked. 🥳

- However, I don't know how to get the example/code snippet for `Pixels` and `Degrees` to be rendered in the docs too since these are 'just' variables and not classes/functions.
- Also, is there a way to remove the trailing `Methods` header [on the Percent wiki page](https://manimce--2604.org.readthedocs.build/en/2604/reference/manim.utils.unit.Percent.html#manim.utils.unit.Percent) since this isn't really applicable?

File: manim/utils/unit.py
Comment (behackl) [inline manim/utils/unit.py:28]:
```suggestion
        ``Z_AXIS`` has no such frame limits so cannot be used here.
```

File: manim/utils/unit.py
Comment (behackl) [inline manim/utils/unit.py:98]: This should probably still be in an "Examples" section of the docstring.

File: manim/utils/unit.py
Comment (behackl) [inline manim/utils/unit.py:89]: This comment is a bit problematic in so far, as this depends on the selected render quality -- and we use rather particular setting for the documentation that is actually different from all of the other presets (`medium_quality`, `high_quality` etc) that people usually use. Not sure whether there is a good solution that works both in a rendered form in the documentation and with our usual presets. 🤔

**Review by behackl — COMMENTED** [review]
Thank you for your contribution, this is great! I've left some minor comments, and have at least partial answers to your questions:


> how to get the example/code snippet for Pixels and Degrees to be rendered in the docs too

Good question, and I'm actually not sure. A workaround would be to move the descriptions and examples to the module docstring.

> is there a way to remove the trailing Methods header 

Another good question! This might be caused by the template in `docs/source/_templates/autosummary/class.rst` checking for the "wrong" thing on line 12. I can't offer a workaround right away, but maybe you want to take a look yourself? :-)

**Comment by tameTNT** [comment]
That's great - thank you so much for the reviews! I'll have a look at addressing them properly when I have time over the next few days. 👍

**Comment by tameTNT** [comment]
Just to mark that this PR isn't dead - I will continue with this soon! Unfortunately some personal issues came up over the last week that meant I was unable to do the further work I would have liked.

File: manim/utils/unit.py
Comment (tameTNT) [reply manim/utils/unit.py:89]: My thinking here is that perhaps I could just explicitly state the conversion used by the docs with a disclaimer that this changes based on quality settings.
For example, something similar to:
```python
# With the low quality (-ql) Manim frame width/ratio,
#     150 Pixels = 2.5 Munits
# This ratio will change depending on your resolution, quality configuration etc.
```
Where would I find the documentation's Manim configuration?
I tried doing some digging in `\docs\` and `\docs\source\` while the documentation was building locally but haven't been able to find anything yet.

If the configuration is entirely unwieldy as you suggest (with some weird pixel-munit ratio), I might just remove the manim directive entirely and leave it as a plaintext example.

**Comment by tameTNT** [comment]
> This might be caused by the template in `docs/source/_templates/autosummary/class.rst` checking for the "wrong" thing on line 12.
@behackl 

I think this might due to the fact that `Percent` has special members (in this case also `inherited_members`) of `__mul__` and `__rmul__` so `{%- if methods %}` evaluates to True generating the methods header - `.. rubric:: {{ _('Methods') }}` - but then `for item in methods if item != '__init__' and item not in inherited_members` is an empty list so no methods are actually added beneath the header.

I believe the above change to `class.rst` *might* fix it but my Jinga's not the best 😂

File: manim/utils/unit.py
Comment (tameTNT) [reply manim/utils/unit.py:89]: Alternatively, I could just include the `quality` option (see commit below)
```python
    .. manim :: PixelsExample
        :quality: low
        :ref_classes: Dot ApplyMethod
```

**Comment by tameTNT** [comment]
Currently looking into something with filters (https://jinja.palletsprojects.com/en/3.0.x/templates/#list-of-builtin-filters).

`{%- if methods|select('ne', '__init__')|select('in', inherited_members) %}` for line 12 in `class.rst`.

But we actually need **notin** since we're trying to match the list generated in `{% for item in methods if item != '__init__' and item not in inherited_members %}`.

For now, I don't think I'll be able to do any more of this today but I'll have another look tomorrow. Any suggestions in the meantime are greatly appreciated!

**Comment by tameTNT** [comment]
I've exhausted my options I think at this point. I'm going to open this PR up for review with the hope that someone will know how to address the bug (for initial description see [the earlier comment](https://github.com/ManimCommunity/manim/pull/2604#issuecomment-1076414144)).

To aid them, here's a summary of what I've tried in the commits above:

- `{% set methods_to_list = [item for item in methods if item != '__init__' and item not in inherited_members] %}`
- ```
{% set no_init_m = methods | select('ne', '__init__') | set %}
{%- if no_init_m.difference(set(inherited_members)) %}
...
{% for item in methods if item != '__init__' and item not in inherited_members %}
```
- ```
{% set no_init_m = methods | select('ne', '__init__') %}
{%- if set(no_init_m).difference(set(inherited_members)) %}
```
- `{% set methods_to_list = methods if item != '__init__' and item not in inherited_members %}`
- ```
{% set methods_to_list = methods if item not in inherited_members %}
{% set methods_to_list = methods_to_list if item != '__init__' %}
```
- ```
   {% set displayed_methods = [] %}
   {% for item in methods %}
      {% if item != '__init__' and item not in inherited_members %}
         {% do displayed_methods.append(item) %}
      {%- endif %}
   {%- endfor %}
```
- Above but with `{% set displayed_methods = displayed_methods + [item] %}` instead of do
- ```
   {%- for m in methods %}
      {%- if m != '__init__' %}
         {%- if m not in inherited_members %}
            {% set displayed_methods = displayed_methods + [m] %}
         {%- endif %}
      {%- endif %}
   {%- endfor %}
```

I hope someone will be able to narrow down what I haven't covered yet. Currently, no methods are displayed at all on any pages (check the links in the starter comment to compare stable docs with PR docs).

**Comment by kilacoda** [comment]
Hi, just checking up on this. Is it ready for review? You've marked it as a draft and have requested a review from @behackl  so that's why it seems a bit confusing. The changes seem fine to me tbh so if you think this is ready for review I'd be up for reviewing and merging it soon.

**Comment by tameTNT** [comment]
Hi! Yeah so as my last comment summarises that I'd hit a snag when it came to editing the documentation template that I wasn't sure how to fix. It outlines what I've already tried to fix the bug so I'd marked the PR as a draft until someone found a solution (it's probably simple I just don't know Jenja well enough 😂). But in terms of the actual documentation this PR _is_ complete. 👍

**Review by chopan050 — APPROVED** [review]
Hello!
I'm one of the devs working on the PR #3484 - me and JasonGrace are doing something related to what you need regarding the docs.
Mainly, we made our custom Sphinx directive and file parsing to **display docs for Manim's type aliases (which are also module variables)**, and currently we're trying to make hyperlinks to those new types.

What might interest you is **how we parsed the docstrings for module variables.** We also had issues with multi-line docs and how they got messed up with colons, dots and empty lines. In the end, what we did is **read the file contents of the module instead of importing it, and parsing them with `ast` to extract desired info.** You might wanna take a look at that.

You can take a look at my PR and read the new `manim/utils/docbuild/module_parsing.py` file there for some inspiration.

Otherwise, I'd suggest you to wait until my PR is done and merged for your two problems to get fixed.

---

In the meantime, I think that your contribution is nice! **LGTM! I approve it as it is now.** I believe that we could figure out the details and fixes for another PR.

File: manim/utils/unit.py
Comment (JasonGrace2282) [inline manim/utils/unit.py:5]:
```suggestion
```
There's no need to add this, it just creates visual clutter.

File: manim/utils/unit.py
Comment (JasonGrace2282) [inline manim/utils/unit.py:103]:
```suggestion
"""A scaling object to convert pixels to Munits."""
```

File: manim/utils/unit.py
Comment (JasonGrace2282) [inline manim/utils/unit.py:106]:
```suggestion
"""A convenience constant to convert degrees to radians."""
```

**Review by JasonGrace2282 — CHANGES_REQUESTED** [review]
Looks good overall! Just some small changes would be nice :)

**Comment by tameTNT** [comment]
Hi! Thank you for somehow finding this old relic of a PR 😆 and for contributing to the project 🤩.

I've just taken some time to refamiliarise myself with what I was even doing here and some observations:

- The removal of the basic Pixels example as suggested by @JasonGrace2282 is fair I think, but maybe a small example line should be included in the (as with `Unit.Degrees`) so there is at least some sample usage even if there isn't a full video.
- I can see the work you've done in your PR and it looks cool! As you say, as you've clearly reworked a large part of this part of the docs rendering process, best to wait until your PR is merged to transition the three module attribute docs (Pixels, Degrees, MUnits) for this PR to be rendered in the docs using your new tools (and then put the examples in line instead of under separate headings as was compromised initially here).
- The methods heading changes I've made (in `docs/source/_templates/autosummary/class.rst`) should be rolled back before any merging with this branch takes place as these changes currently break existing documentation pages (e.g. **`Methods`** header missing on `TexTemplate` page on [this PR's build](https://manimce--2604.org.readthedocs.build/en/2604/reference/manim.utils.tex.TexTemplate.html#textemplate) (c.f. [stable docs](https://docs.manim.community/en/stable/reference/manim.utils.tex.TexTemplate.html#:~:text=str-,METHODS,-add_to_document))). Once your PR is merged then this can be reattempted of course in a separate PR is probably best with a more appropriate name, etc.

**Comment by JasonGrace2282** [comment]
If you're still interested, Chopan's PR has now been merged.

**Comment by tameTNT** [comment]
To be completely honest, I don't really have the time at the moment nor the familiarity with the project/repo as a whole anymore since it has been so long since I worked on it, so I don't think I'd be able to do a satisfactory job.

**Comment by JasonGrace2282** [comment]
That's completely fine, life comes first! I'll just close this PR and create an issue to remind ourselves about this.

**Pull request closed without merging**
